docs/source/plugin/what.rst

```diff
@@ -0,0 +1,133 @@
+=================
+What is a Plugin?
+=================
+
+Before writing your first plugin, you may want to know what is a plugin, how is
+it made, and what are the best solutions for your need. In this chapter we'll
+cover what is a Plugin, what you can do with them, and some vocabulary.
+
+You may want to skip ahead to :doc:`anatomy` if you already know what a Plugin
+is and what types of plugin there are.
+
+.. contents::
+    :local:
+    :depth: 2
+
+Plugin Types
+============
+
+There are four types of plugins: :term:`Single file plugin`,
+:term:`Folder plugin`, :term:`Namespace package plugin`, and
+:term:`Entry point plugin`.
+
+Single file
+-----------
+
+A **Single file** plugin is the most basic form a Sopel plugin can take. It is
+composed of a single file available to Sopel from its plugin's directory (or
+from one of its
+:attr:`extra directories <sopel.config.core_section.CoreSection.extra>`). You
+should write a single file plugin when:
+
+* everything is in a single file
+* you don't have any dependencies other than Sopel and the Python's built-in
+  library
```

Comment:
```suggestion
* you don't have any dependencies other than Sopel and Python's standard library
```

Aside: I don't entirely agree with this—it's perfectly fine to write a single-file plugin with third-party dependencies, in my mind, especially if it's not intended for distribution.

I might add more suggestions about this in the next review round, if you don't vehemently object to explicitly giving users the freedom to skip learning about entry points and packages right away, while they ease into plugin (or Python) development first.

Reply:
Well it's complicated. Probably worth a conversation.

Reply:
New alternative suggestion, since we haven't gotten around to having this conversation yet… (sadly)

```suggestion
* you don't have any dependencies other than Sopel and Python's built-in library,
  or your plugin is a "one-off" for your use only (not intended to be shared)
```

I still don't see why we'd mention dependencies for this, but not the `Folder` type.

Reply:
Oh we totally should @ folder plugins!

Reply:
Going to do that during rebase before final review?